ci-operator/step-registry/gather/gcp-console: Gather console logs on GCP

wking opened the pull request:
Like the similar aws-console step, but for GCP, following [these docs][1].  I'm not sure why they require a `--zone` argument when `instances list` seems to be able to figure this out, but whatever.  Leaving it off doesn't seem to be an option:

```console
wking@cloudshell:~ (openshift-gce-devel-ci)$ gcloud --format json compute instances get-serial-port-output "${INSTANCE_ID}" </dev/null >logs.json
ERROR: (gcloud.compute.instances.get-serial-port-output) Underspecified resource [ci-ln-2gtd6db-f76d1-fj4cl-master-0]. Specify the [--zone] flag.
wking@cloudshell:~ (openshift-gce-devel-ci)$ gcloud --version | head -n1
Google Cloud SDK 315.0.0
```

`OWNERS` are like AWS, from 390d6bee20 (#9743), and also [the installer's GCP approvers][2].  CC @cgwalters, @enxebre, @jstuever, @patrickdillon, @vrutkovs.

[1]: https://cloud.google.com/compute/docs/reference/rest/v1/instances/getSerialPortOutput
[2]: https://github.com/openshift/installer/blob/e17697302ff991ae80d6e9c0c10426db510881e2/OWNERS_ALIASES#L60-L62

abhinavdahiya commented on ci-operator/step-registry/gather/gcp-console/gather-gcp-console-commands.sh:
```diff
@@ -0,0 +1,50 @@
+	wait "$!"
+else
+	echo "No kubeconfig; skipping providerID extraction."
+	exit 0
```
This would prevent collection from shared_dir gce instance ids right below

wking replied:
Fixed with 64c641c88e -> 32c330271b.

wking replied:
And I've filed #14401 to fix this same issue in the AWS step.

abhinavdahiya commented on ci-operator/step-registry/gather/gcp-console/gather-gcp-console-commands.sh:
```diff
@@ -0,0 +1,50 @@
+	oc --request-timeout=5s get nodes -o jsonpath --template '{range .items[*]}{.spec.providerID}{"\n"}{end}' | sed 's|.*/||' > "${TMPDIR}/node-provider-IDs.txt" &
+	wait "$!"
+
+	oc --request-timeout=5s -n openshift-machine-api get machines -o jsonpath --template '{range .items[*]}{.spec.providerID}{"\n"}{end}' | sed 's|.*/||' >> "${TMPDIR}/node-provider-IDs" &
```
```suggestion
	oc --request-timeout=5s -n openshift-machine-api get machines -o jsonpath --template '{range .items[*]}{.spec.providerID}{"\n"}{end}' | sed 's|.*/||' >> "${TMPDIR}/node-provider-IDs.txt" &
```

wking replied:
Fixed with 64c641c88e -> 32c330271b.

wking replied:
#14396 fixed this same issue in the AWS step.

wking commented on ci-operator/step-registry/gather/gcp-console/gather-gcp-console-commands.sh:
```diff
@@ -0,0 +1,48 @@
+cat "${TMPDIR}/node-provider-IDs.txt" | sort | uniq | while read -r INSTANCE_ID
+do
+	echo "Finding the zone for ${INSTANCE_ID}"
+	ZONE="$(gcloud --format json compute instances list "--filter=name=(${INSTANCE_ID})" | jq -r '.[].zone')" &
+	wait "$!"
```
Hrm, I don't understand [this][1] (from [this job][2]):

```
Finding the zone for ci-op-yjsrfl7s-d1016-tpmpm-master-0
/bin/bash: line 44: ZONE: unbound variable
error: failed to execute wrapped command: exit status 1
```

Because I'm trying to set that variable a few lines up:

```console
$ git grep -2hn ZONE ci-operator/step-registry/gather/gcp-console/gather-gcp-console-commands.sh | head -n8
37-do
38-     echo "Finding the zone for ${INSTANCE_ID}"
39:     ZONE="$(gcloud --format json compute instances list "--filter=name=(${INSTANCE_ID})" | jq -r '.[].zone')" &
40-     wait "$!"
41:     if test -z "${ZONE}"
42-     then
43-             echo "No zone found for ${INSTANCE_ID}, so not attempting to gather console logs"
44-     fi
```

Maybe I can't background the variable-set.

[1]: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/13398/rehearse-13398-pull-ci-openshift-cluster-etcd-operator-release-4.6-e2e-agnostic/1339343972186722304/artifacts/e2e-agnostic/gather-gcp-console/container-logs/test.log
[2]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/13398/rehearse-13398-pull-ci-openshift-cluster-etcd-operator-release-4.6-e2e-agnostic/1339343972186722304

wking replied:
I've pushed d4bb001b65 -> 5d9d973fbe to move the `wait` inside the subshell.

cgwalters reviewed (approved):
Looks plausible to me.

cgwalters commented:
I don't think this worked, see errors in this job https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-ocp-installer-e2e-gcp-rt-4.7/1339905732908158976
```
 Dec 18 12:56:53.204: INFO: lookupDiskImageSources: gcloud error with [[]string{"instance-groups", "list-instances", "", "--format=get(instance)"}]; err:exit status 1
Dec 18 12:56:53.204: INFO:  > ERROR: (gcloud.compute.instance-groups.list-instances) could not parse resource []
Dec 18 12:56:53.204: INFO:  > 
```

wking commented:
> ...see errors in this job https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-ocp-installer-e2e-gcp-rt-4.7/1339905732908158976 ...

That's an old-style release-template job (no per-step directories [here][1]), so this step wasn't involved at all there.  But yeah, seems to be having some trouble.  See [this untrusted API-server after a failed install][2]:

```
2020/12/19 21:41:53 Container test in pod e2e-gcp-console-ipi-install-install failed, exit code 1, reason Error
2020/12/19 21:41:54 Executing pod "e2e-gcp-console-gather-gcp-console"
2020/12/19 21:41:59 Container cp-secret-wrapper in pod e2e-gcp-console-gather-gcp-console completed successfully
Activated service account credentials for: [do-not-delete-ci-provisioner@openshift-gce-devel-ci.iam.gserviceaccount.com]
Updated property [core/project].
Unable to connect to the server: x509: certificate signed by unknown authority (possibly because of "crypto/rsa: verification error" while trying to verify candidate authority certificate "kube-apiserver-lb-signer")
error: failed to execute wrapped command: exit status 1
2020/12/19 21:42:04 Container test in pod e2e-gcp-console-gather-gcp-console failed, exit code 1, reason Error
```

And [this failure to find a zone after a failed bootstrap][3]:

```
2020/12/19 12:29:58 Executing pod "e2e-gcp-upgrade-gather-gcp-console"
2020/12/19 12:30:03 Container cp-secret-wrapper in pod e2e-gcp-upgrade-gather-gcp-console completed successfully
Activated service account credentials for: [do-not-delete-ci-provisioner@openshift-gce-devel-ci.iam.gserviceaccount.com]
Updated property [core/project].
Finding the zone for 
No zone found for , so not attempting to gather console logs
Gathering console logs for  from 
ERROR: (gcloud.compute.instances.get-serial-port-output) could not parse resource []
error: failed to execute wrapped command: exit status 1
```

I'll get a fixup up for these cases where the cluster-install had trouble...

[1]: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/release-openshift-ocp-installer-e2e-gcp-rt-4.7/1339905732908158976/artifacts/e2e-gcp/
[2]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_console/7600/pull-ci-openshift-console-master-e2e-gcp-console/1340411780635037696#1:build-log.txt%3A66
[3]: https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-e2e-gcp-upgrade-4.7/1340262366092201984#1:build-log.txt%3A171

wking commented:
On the other hand [here][1] is a job where the current step [worked][2].

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/promote-release-openshift-okd-machine-os-content-e2e-gcp-4.7/1340247868513980416
[2]: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/promote-release-openshift-okd-machine-os-content-e2e-gcp-4.7/1340247868513980416/artifacts/e2e-gcp/gather-gcp-console/
